File: src/main/java/com/multicampus/gamesungcoding/a11ymarketserver/common/jwt/provider/JwtTokenProvider.java
package com.multicampus.gamesungcoding.a11ymarketserver.common.jwt.provider;

import com.multicampus.gamesungcoding.a11ymarketserver.common.properties.JwtProperties;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.MalformedJwtException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.User;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.util.Arrays;
import java.util.Collection;
import java.util.Date;
import java.util.stream.Collectors;

@Slf4j
@Component
public class JwtTokenProvider {
    private final SecretKey secretKey;
    private final long accessTokenValidityMs;

    public JwtTokenProvider(JwtProperties jwtProperties) {
        this.secretKey = jwtProperties.getSecretKey();
        this.accessTokenValidityMs = jwtProperties.getAccessTokenValidityMs();
    }

    public String createAccessToken(Authentication authentication) {
        // e.g., "ROLE_USER,ROLE_ADMIN"
        String authorities = authentication
                .getAuthorities()
                .stream()
                .map(GrantedAuthority::getAuthority)
                .collect(Collectors.joining(","));

        var now = new Date();
        var expiryDate = new Date(now.getTime() + accessTokenValidityMs);

        return Jwts.builder()
                .subject(authentication.getName())
                .claim("auth", authorities)
                .issuedAt(now)
                .expiration(expiryDate)
                .signWith(secretKey)
                .compact();
    }

    public Authentication getAuthentication(String token) {
        var claims = Jwts.parser()
                .verifyWith(secretKey)
                .build()
                .parseSignedClaims(token)
                .getPayload();

        Collection<? extends GrantedAuthority> authorities =
                Arrays.stream(claims.get("auth").toString().split(","))
                        .map(SimpleGrantedAuthority::new)
                        .toList();
        var principal = new User(claims.getSubject(), "", authorities);
        return new UsernamePasswordAuthenticationToken(principal, token, principal.getAuthorities());
    }

    public boolean validateToken(String token) {
        try {
            log.debug("JwtTokenProvider - validateToken");
            Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token);
            return true;
        } catch (MalformedJwtException e) {
            log.debug("JwtTokenProvider.getAuthentication - Invalid JWT token format: {}", e.getMessage());
        } catch (ExpiredJwtException e) {
            log.debug("JwtTokenProvider.getAuthentication - Expired JWT token: {}", e.getMessage());
        } catch (SecurityException e) {
            log.debug("JwtTokenProvider.getAuthentication - Invalid JWT signature: {}", e.getMessage());
        } catch (Exception e) {
            log.debug("JwtTokenProvider.getAuthentication - JWT token validation error: {}", e.getMessage());
        }
        return false;
    }
}

File: src/main/kotlin/com/multicampus/gamesungcoding/a11ymarketserver/common/jwt/provider/JwtTokenProvider.kt
package com.multicampus.gamesungcoding.a11ymarketserver.common.jwt.provider

import com.multicampus.gamesungcoding.a11ymarketserver.common.properties.JwtProperties
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.UserRole
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.MalformedJwtException
import org.slf4j.LoggerFactory
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.stereotype.Component
import java.util.*
import javax.crypto.SecretKey

@Component
class JwtTokenProvider(
    jwtProperties: JwtProperties,
) {
    private val log = LoggerFactory.getLogger(this::class.java)

    private val secretKey: SecretKey = jwtProperties.secretKey
    private val accessTokenValidityMs: Long = jwtProperties.accessTokenValidityMs

    fun createAccessToken(authentication: Authentication): String {
        // e.g., "ROLE_USER,ROLE_ADMIN"
        val authorities = authentication
            .authorities
            .joinToString(",") { it.authority }

        return createToken(authentication.name, authorities)
    }

    fun createAccessToken(userId: UUID, role: UserRole): String {
        return createToken(userId.toString(), role.name)
    }

    fun createTemporaryAccessToken(uuid: UUID): String {
        val now = Date()
        val expiryDate = Date(now.time + 15 * 60 * 1000) // 15 minutes

        return Jwts.builder()
            .subject(uuid.toString())
            .claim("auth", UserRole.TEMP.name)
            .issuedAt(now)
            .expiration(expiryDate)
            .signWith(secretKey)
            .compact()
    }

    fun getAuthentication(token: String?): Authentication {
        val claims = Jwts.parser()
            .verifyWith(secretKey)
            .build()
            .parseSignedClaims(token)
            .payload

        val authClaim = claims["auth"]?.toString() ?: ""
        val authorities = authClaim.split(",")
            .filter { it.isNotBlank() }
            .map { SimpleGrantedAuthority(it) }

        val principal = User(claims.subject, "", authorities)
        return UsernamePasswordAuthenticationToken(principal, token, principal.authorities)
    }

    fun validateToken(token: String?): Boolean {
        return try {
            Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token)
            log.debug("JwtTokenProvider - validateToken")
            true
        } catch (e: MalformedJwtException) {
            log.debug("JwtTokenProvider.getAuthentication - Invalid JWT token format: ${e.message}")
            false
        } catch (e: ExpiredJwtException) {
            log.debug("JwtTokenProvider.getAuthentication - Expired JWT token: ${e.message}")
            false
        } catch (e: SecurityException) {
            log.debug("JwtTokenProvider.getAuthentication - Invalid JWT signature: ${e.message}")
            false
        } catch (e: Exception) {
            log.debug("JwtTokenProvider.getAuthentication - JWT token validation error: ${e.message}")
            false
        }
    }

    private fun createToken(subject: String, role: String): String {
        val now = Date()
        val expiryDate = Date(now.time + accessTokenValidityMs)

        return Jwts.builder()
            .subject(subject)
            .claim("auth", role)
            .issuedAt(now)
            .expiration(expiryDate)
            .signWith(secretKey)
            .compact()
    }
}
